graph/network/diffusion.go

```
@@ -0,0 +1,188 @@
+		v.SubVec(lastV, &tmp)
+		if normDiff(heat, last) < tol {
+			ok = true
+			break
+		}
+		iters--
+		if iters < 0 {
+			break
+		}
+	}
+
+	for i, n := range v.RawVector().Data {
+		h[by.Nodes[i].ID()] = n
+	}
+	return h, ok
+}
+
+// Laplacian is a graph Laplacian matrix.
+type Laplacian struct {
+	// Matrix holds the Laplacian matrix.
+	mat.Matrix
+
+	// Nodes holds the input graph nodes.
+	Nodes []graph.Node
+
+	// Index is a mapping from the graph
+	// node IDs to row and column indices.
+	Index map[int64]int
+}
+
+// NewLaplacian returns a Laplacian matrix for the simple undirected graph g.
+// The Laplacian is defined by D-A where D is a diagonal matrix holding the
+// degree of each node and A is the graph adjacency matrix of the input graph.
+func NewLaplacian(g graph.Undirected) Laplacian {
+	nodes := g.Nodes()
+	indexOf := make(map[int64]int, len(nodes))
+	for i, n := range nodes {
+		id := n.ID()
+		indexOf[id] = i
+	}
+
+	l := mat.NewSymDense(len(nodes), nil)
+	for j, u := range nodes {
+		to := g.From(u)
+		l.SetSym(j, j, float64(len(to)))
+		uid := u.ID()
+		for _, v := range to {
+			vid := v.ID()
+			if uid < vid {
+				l.SetSym(indexOf[vid], j, -1)
+			}
+		}
+	}
+
+	return Laplacian{Matrix: l, Nodes: nodes, Index: indexOf}
+}
+
+// NewSymNormLaplacian returns a symmetric normalized Laplacian matrix for the
+// simple undirected graph g.
+// The Laplacian is defined by I-D^(-1/2)AD^(-1/2) where D is a diagonal matrix holding the
+// degree of each node and A is the graph adjacency matrix of the input graph.
+func NewSymNormLaplacian(g graph.Undirected) Laplacian {
+	nodes := g.Nodes()
+	indexOf := make(map[int64]int, len(nodes))
+	for i, n := range nodes {
+		id := n.ID()
+		indexOf[id] = i
+	}
+
+	l := mat.NewSymDense(len(nodes), nil)
+	for j, u := range nodes {
+		to := g.From(u)
+		if len(to) == 0 {
+			continue
+		}
+		l.SetSym(j, j, 1)
+		uid := u.ID()
+		udeg := math.Sqrt(float64(len(to)))
+		for _, v := range to {
```

Just checking there are no self-loop issues here with overwriting the `Set`

---

The Laplacians cannot be constructed from graphs with self loops.

---

Should there be a check rather than just failing silently?  There's no comment that `Undirected` cannot have self loops.

---

Yeah, we used to refuse self-loops in all (simple) graphs. The three approaches to take are 1) add an `error` return, 2) panic or 3) just document that Laplacians are defined on adjacency matrices with zero diagonals. I don't know.

---

I think panic is fine.

---

Done.